### PR TITLE
Added Redis command ApproxMatch and edited TRIE.INSERT so it can take more than one string

### PR DESCRIPTION
**Pull request by DustinoZhang**
The last of Spellcheck team's requests!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -157,20 +154,54 @@ Your module should be loaded up. Check your server log for an indication that th
...
-       redis> T.INSERT key1 helloworld
-       "Success"
-       redis> T.CONTAINS key1 helloworld
-       "The trie contains the word."
-       redis> T.CONTAINS key1 goodbye
-       "The trie does not contain the word."
-       redis> T.CONTAINS key1 hello
-       "The trie contains it as a prefix but not as a word."
\ No newline at end of file
+TRIE.CONTAINS checks if a string exists in a given trie key. If the key does not previously exist, an error will be thrown. Otherwise, different messages will be printed based on whether the string is contained within the trie. If the string is contained within the trie, the integer 1 will be returned. If the string is not contained within the trie, the integer 0 will be returned. If the string is contained within the trie as a prefix but not as a word, the integer -1 will be returned.
```
[the remark this answers is not in the document]
Comment (DustinoZhang) [reply README.md]: The only way to do that would be to actually return a string, which isn't optimal on the API side. I thought about it but I think returning an integer is the best way to do this.

File: module/trie.c [hunk 2]
```diff
@@ -5,18 +5,22 @@
 #define IN_TRIE 1
 #define NOT_IN_TRIE 0 
 #define PARTIAL_IN_TRIE (-1)
+// The maximum length of a string we're willing to process
+// Longest word in english dictionary is 45 letters
```
[the remark this answers is not in the document]
Comment (dwahme) [reply module/trie.c]: Updated.

File: module/trie.c [hunk 3]
```diff
@@ -269,14 +306,529 @@ int trie_count_completion(struct trie *t, char *pre)
     return trie_count_completion_recursive(end);
 }
 
+/* 
+    Since modules do not include header files typically, this is an early 
+    declaration of the suggestions() function since helper functions use them
```
[the remark this answers is not in the document]
Comment (dwahme) [reply module/trie.c]: Updated.

File: module/trie.c [hunk 4]
```diff
@@ -269,14 +306,529 @@ int trie_count_completion(struct trie *t, char *pre)
...
+ */
+int suggestions(match_t **set, struct trie *t, char *prefix, char *suffix, int edits_left, int n);
+
+/*
+    Sees if there are any words in a trie that contain a given prefix
+    NOTE- awaiting the actual function from support tools
```
[the remark this answers is not in the document]
Comment (dwahme) [reply module/trie.c]: Outdated comment from an old version. Removed.

File: module/trie.c [hunk 5]
```diff
@@ -269,14 +306,529 @@ int trie_count_completion(struct trie *t, char *pre)
...
+    int rc = EXIT_SUCCESS;
+    int len = strlen(prefix);
+    char* new_prefix;
+
+    // Ran out of space
+    if (len == MAXLEN - 1) {
+        return EXIT_FAILURE;
+    }
+
+    for (i = 1; i < 248; i++) {
```
[the remark this answers is not in the document]
Comment (dwahme) [reply module/trie.c]: suggestions() crashes with extended ASCII codes 248-255. I'm not sure why this happens; it's specific to the values themselves.

File: module/trie.c [hunk 6]
```diff
@@ -269,14 +306,529 @@ int trie_count_completion(struct trie *t, char *pre)
...
+                }
+            }
+        }
+    }
+
+    return EXIT_SUCCESS;
+}
+
+/*
+    Performs a mass approximate sting match on all the words in a dictionary beginning with a prefix
```
[the remark this answers is not in the document]
Comment (dwahme) [reply module/trie.c]: Fixed.

File: module/trie.c [hunk 7]
```diff
@@ -269,14 +306,529 @@ int trie_count_completion(struct trie *t, char *pre)
...
+    For more information on approx matching: https://www.wikiwand.com/en/Approximate_string_matching
+  
+    Parameters:
+     - set: An array of match_t*'s
+     - t: A trie. Must point to a trie allocated with trie_new
+     - prefix: A prefix of a word. Must be contained in a dictionary or ""
+     - suffix: A suffix of a word. Can also be ""
+     - edits_left: The maximum levenshtein distance a word can be from prefix+suffix
+     - n: The length of set (amount of matches in the set)
```
[the remark this answers is not in the document]
Comment (dwahme) [reply module/trie.c]: Fixed

File: module/trie.c [hunk 8]
```diff
@@ -269,14 +306,529 @@ int trie_count_completion(struct trie *t, char *pre)
...
+     - str: A string. This will be the (misspelled) word to match
+     - max_edits: the maximum levenshtein distance the words in the set can have
+     - n: the number of strings to return. 
+  
+    Returns:
+     - The first n strings with the smallest distance, where ties are broken by alphabetical order.
+       If there aren't enough matching strings, each remaining spot is set to NULL.
+     - NULL if there was an error
+ */
+char** suggestion_list(struct trie *t, char *str, int max_edits, int amount) {
```
[the remark this answers is not in the document]
Comment (dwahme) [reply module/trie.c]: Fixed.

File: README.md [hunk 9]
```diff
@@ -16,57 +16,54 @@ The trie class is composed of the trie_t struct and its respective operations.
...
     **Purpose:** Creates new node in trie_t.
     
     **Details:** Set t->children[current] to be current, is_word for new node set to 0. 
 
-3. int insert_string(char \*word, trie_t \*t) 
+3. int trie_insert_string(trie_t \*t, char \*word) 
 
     **Purpose:** Inserts word into trie.
 
     **Details:** For each trie, check if entry of the next character exists in the children array:
```
[the remark this answers is not in the document]
Comment (dwahme) [reply README.md]: Updated by Dustin

File: README.md [hunk 10]
```diff
@@ -16,57 +16,54 @@ The trie class is composed of the trie_t struct and its respective operations.
...
 
-                 If so, move into that node in the array
+4. int trie_char_exists(trie_t \*t, char c)
 
-                 If not, add a new node
+    **Purpose:** Checks if a char exists in a trie
 
-                 Then move on to the next character in string
+    **Details:** Checks if char c is contained within the trie
```
[the remark this answers is not in the document]
Comment (dwahme) [reply README.md]: Updated by Dustin.

File: README.md [hunk 11]
```diff
@@ -16,57 +16,54 @@ The trie class is composed of the trie_t struct and its respective operations.
...
-                 If word is not in trie, trie is not modified. 
-                   
-                 If word is completely unique (no other part of the word is part of another word) then delete the entire word.
+6. int trie_search(trie_t \*t, char \*word)
 
-                 If word is the prefix of another word in the trie, unmark the leaf node.
+    **Purpose:** Search for a word in a trie. 
 
-                 If word is present in the trie, having at least one other word as a prefix, delete all the nodes up to the prefix.
+    **Details:** Returns 1 if word is found. Returns 0 if word is not found at all and -1 if word is found but end node's is_word is 0.
```
[the remark this answers is not in the document]
Comment (dwahme) [reply README.md]: Updated by Dustin.

File: README.md [hunk 12]
```diff
@@ -16,57 +16,54 @@ The trie class is composed of the trie_t struct and its respective operations.
...
 
-5. int trie_search(char \*word, trie_t \*t)
+7. int trie_count_completion(trie_t \*t, char \*pre)
 
-    **Purpose:** Search for a word in a trie. 
+    **Purpose:** Count the number of different possible endings of a given prefix in a trie 
 
-    **Details:** Returns 1 if word is found. Returns 0 if word is not found at all and -1 if word is found but end node's is_word is 0.
+    **Details:** Returns an integer of the number of endings of the prefix given, 0 if prefix doesn't exist
```
[the remark this answers is not in the document]
Comment (dwahme) [reply README.md]: Updated by Dustin.

File: README.md [hunk 13]
```diff
@@ -157,20 +154,54 @@ Your module should be loaded up. Check your server log for an indication that th
...
+       (int) 0
+       redis> TRIE.CONTAINS key1 helloworld
+       (int) 1
+       redis> TRIE.CONTAINS key1 goodbye
+       (int) 0
+       redis> TRIE.CONTAINS key1 hello
+       (int) -1
+
+### TRIE.APPROXMATCH key value1 (optional)value2 (optional)value3
+TRIE.APPROXMATCH returns a list of suggested words that have the given prefix. It requires at least a key, whose value is an existing trie, and a prefix (value1) to look for within the trie. The first optional argument (value2) specifies the edit distance (or how close the words returned can be to the given prefix). If no value is given, the default is 2. The second optional argument (value 3) specifies the number of "matches", or possible completions, that will be returned by the command. If no value is given, the default is 10 (meaning 10 possible words will be given, if there aren't 10 possible endings the remaining slots will be filled with Redis (nil) values).
```
[the remark this answers is not in the document]
Comment (dwahme) [reply README.md]: Updated to "### TRIE.APPROXMATCH key prefix (optional)max_edit_distance (optional)num_matches".

Second part updated by Dustin.

File: module/trie.c [hunk 14]
```diff
@@ -269,14 +306,529 @@ int trie_count_completion(struct trie *t, char *pre)
...
+    assert(t != NULL);
+    assert(str != NULL);
+
+    match_t **set = suggestion_set_new(t, str, max_edits, amount);
+
+    if (set == NULL) {
+        return NULL;
+    }
+
+    char **results = suggestion_set_first_n(set, amount);
```
[the remark this answers is not in the document]
Comment (dwahme) [reply module/trie.c]: Updated by Dustin

File: module/trie.c [hunk 15]
```diff
@@ -287,12 +839,17 @@ int TrieInsert_RedisCommand(RedisModuleCtx *ctx, RedisModuleString **argv,
         return RedisModule_ReplyWithError(ctx, REDISMODULE_ERRORMSG_WRONGTYPE);
     }
     size_t dummy;
-    char *temp = RedisModule_StringPtrLen(argv[2], &dummy);
+    /* Number of strings wanting to be inserted */
```
[the remark this answers is not in the document]
Comment (dwahme) [reply module/trie.c]: Updated by Dustin

File: module/trie.c [hunk 16]
```diff
@@ -336,17 +898,71 @@ int TrieContains_RedisCommand(RedisModuleCtx *ctx, RedisModuleString **argv,
...
-        RedisModule_ReplyWithSimpleString(ctx, "The trie does not contain the word.");
-    } else {
-        RedisModule_ReplyWithSimpleString(ctx, "The trie contains it as a prefix but not as a word.");  
-    }       
+    RedisModule_ReplyWithLongLong(ctx, c);      
     RedisModule_ReplicateVerbatim(ctx);
     return REDISMODULE_OK;
 }
 
+/* TRIE.APPROXMATCH key value (opt)value (opt)value */
```
[the remark this answers is not in the document]
Comment (dwahme) [reply module/trie.c]: Updated to `TRIE.APPROXMATCH key prefix (optional)max_edit_distance (optional)num_matches`

File: module/trie.c [hunk 17]
```diff
@@ -336,17 +898,71 @@ int TrieContains_RedisCommand(RedisModuleCtx *ctx, RedisModuleString **argv,
...
+        return RedisModule_ReplyWithError(ctx, "ERR invalid key: not an existing trie");
+    }
+    else if (RedisModule_ModuleTypeGetType(key) != trie)
+    {
+        return RedisModule_ReplyWithError(ctx, REDISMODULE_ERRORMSG_WRONGTYPE);
+    }
+    
+    size_t dummy;
+    char *temp = RedisModule_StringPtrLen(argv[2], &dummy);
+    long long medits = 2;
```
[the remark this answers is not in the document]
Comment (dwahme) [reply module/trie.c]: Comment added by Dustin.

File: README.md [hunk 18]
```diff
@@ -35,33 +35,33 @@ The Operations for trie_t are as follows:
 
 3. int trie_insert_string(trie_t \*t, char \*word) 
 
-    **Purpose:** Inserts word into trie.
+    **Purpose:** Inserts word into trie. 
 
-    **Details:** For each trie, check if entry of the next character exists in the children array:
+    **Details:** Returns 0 if word is successfully inserted into trie, 1 if otherwise. In the trie containing the last character, is_word is set to 1.
```
Comment (valeriehan6) [inline README.md]: Typo. "1 otherwise" instead of "1 if otherwise"

**Comment by valeriehan6** [comment]
@mpcsh Ready for review. (Note that I accidentally changed src/trie.c instead of modules/trie.c but have since reverted it back to the last PR version.)

File: README.md [hunk 19]
```diff
@@ -16,57 +16,54 @@ The trie class is composed of the trie_t struct and its respective operations.
 The trie_t struct is composed of:
```
Comment (mpcsh) [inline README.md]: In the interest of time I'm not going to read this README now. We'll do a README review closer to the end of the project, unless you'd like me to do it now.

File: module/trie.c [hunk 20]
```diff
@@ -170,6 +183,30 @@ int trie_insert_string(struct trie *t, char *word)
...
+    Checks if a char exists in a trie 
+    Parameters:
+     - t: A pointer to the given trie
+     - c: The character we want to check
+
+    Returns:
+     - EXIT_SUCCESS if c exists in t
+     - EXIT_FAILURE if it doesn't
+*/
+int trie_char_exists(struct trie *t, char c)
```
Comment (mpcsh) [inline module/trie.c]: Let's have this return a `bool`.

File: module/trie.c [hunk 21]
```diff
@@ -170,6 +183,30 @@ int trie_insert_string(struct trie *t, char *word)
...
+     - EXIT_FAILURE if it doesn't
+*/
+int trie_char_exists(struct trie *t, char c) 
+{
+    assert(t != NULL);
+    assert(t->charlist != NULL);
+
+    int index = (int)c;
+
+    if (t->charlist[index] == '\0') {
```
Comment (mpcsh) [inline module/trie.c]: By having this function return a `bool`, we can condense these four lines into `return t->charlist[index] == '\0'`

File: module/trie.c [hunk 22]
```diff
@@ -269,14 +306,528 @@ int trie_count_completion(struct trie *t, char *pre)
...
+  
+    Parameters: 
+     - t: A trie. Must point to a trie allocated with trie_new
+     - s: A string. The prefix to check
+  
+    Returns:
+     - true if there is a word in the dictionary with the prefix
+     - false otherwise
+*/
+bool has_children(struct trie *t, char *s)
```
Comment (mpcsh) [inline module/trie.c]: Let's rename this to `trie_has_children`.

File: module/trie.c [hunk 23]
```diff
@@ -269,14 +306,528 @@ int trie_count_completion(struct trie *t, char *pre)
...
+     - t: A trie. Must point to a trie allocated with trie_new
+     - s: A string. The prefix to check
+  
+    Returns:
+     - true if there is a word in the dictionary with the prefix
+     - false otherwise
+*/
+bool has_children(struct trie *t, char *s) 
+{    
+    if (trie_search(t, s) == NOT_IN_TRIE) {
```
Comment (mpcsh) [inline module/trie.c]: Why not just `return trie_search(t, s) == NOT_IN_TRIE`?

File: module/trie.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (mpcsh) [reply module/trie.c]: Let's document this in a comment above the loop.

File: README.md (hunk 19 above)
Comment (DustinoZhang) [reply README.md]: That's fine!

File: module/trie.c (hunk 23 above)
Comment (dwahme) [reply module/trie.c]: That is a lot more elegant, good catch.

**Comment by DustinoZhang** [comment]
@mpcsh Ready for review again

**Pull request merged**